File: src/client/java/fr/hardel/asset_editor/client/javafx/components/page/enchantment/ExclusiveGroupSection.java
package fr.hardel.asset_editor.client.javafx.components.page.enchantment;

import fr.hardel.asset_editor.AssetEditor;
import fr.hardel.asset_editor.client.javafx.VoxelColors;
import fr.hardel.asset_editor.client.javafx.VoxelFonts;
import fr.hardel.asset_editor.client.javafx.components.ui.ResponsiveGrid;
import fr.hardel.asset_editor.client.javafx.components.ui.Category;
import fr.hardel.asset_editor.client.javafx.lib.StudioContext;
import fr.hardel.asset_editor.store.EnchantmentFlushAdapter;
import fr.hardel.asset_editor.client.javafx.lib.data.ExclusiveSetGroup;
import fr.hardel.asset_editor.client.javafx.lib.data.StudioBreakpoint;
import fr.hardel.asset_editor.client.javafx.lib.StudioText;
import fr.hardel.asset_editor.workspace.action.EditorAction;
import javafx.geometry.Insets;
import javafx.scene.control.Label;
import javafx.scene.layout.VBox;
import net.minecraft.client.resources.language.I18n;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.Identifier;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.function.Function;

public final class ExclusiveGroupSection extends VBox {

    private final StudioContext context;
    private final Identifier elementId;
    private final Function<EditorAction, Boolean> dispatchAction;
    private final Function<Identifier, Boolean> applyTag;
    private EnchantmentTags currentTargetCard;

    public ExclusiveGroupSection(StudioContext context, Identifier elementId,
        Function<EditorAction, Boolean> dispatchAction,
        Function<Identifier, Boolean> applyTag) {
        this.context = context;
        this.elementId = elementId;
        this.dispatchAction = dispatchAction;
        this.applyTag = applyTag;
        setSpacing(32);
        setMaxWidth(Double.MAX_VALUE);

        var entry = context.currentEntry(Registries.ENCHANTMENT);
        String currentExclusiveTag = "";
        Set<Identifier> currentTags = Set.of();
        if (entry != null) {
            currentExclusiveTag = entry.data().exclusiveSet().unwrapKey()
                .map(k -> k.location().toString()).orElse("");
            currentTags = entry.tags();
        }

        Function<String, String> labelResolver = value -> {
            if (value == null || value.isBlank())
                return "";
            Identifier id = Identifier.tryParse(value.startsWith("#") ? value.substring(1) : value);
            if (id == null)
                return value;
            return context.resolveHolder(Registries.ENCHANTMENT, id)
                .map(holder -> holder.value().description().getString())
                .orElseGet(() -> StudioText.resolve(Registries.ENCHANTMENT, id));
        };

        getChildren().addAll(
            buildVanillaCategory(currentExclusiveTag, currentTags, labelResolver),
            buildCustomCategory(currentExclusiveTag, currentTags, labelResolver));
    }

    private Category buildVanillaCategory(String currentExclusiveTag, Set<Identifier> currentTags,
        Function<String, String> labelResolver) {
        Category category = new Category(I18n.get("enchantment.exclusive:vanilla"));

        ResponsiveGrid grid = new ResponsiveGrid(ResponsiveGrid.autoFit(256))
            .atMost(StudioBreakpoint.XL, ResponsiveGrid.fixed(1));

        for (ExclusiveSetGroup group : ExclusiveSetGroup.ALL) {
            Identifier tagId = group.tagId();
            String rawTag = tagId.toString();
            EnchantmentTags card = buildCard(
                StudioText.resolve("enchantment_tag", tagId),
                I18n.get("enchantment_tag:" + tagId + ".desc"),
                group.image(), tagId, rawTag,
                currentExclusiveTag, currentTags, labelResolver);
            if (rawTag.equals(currentExclusiveTag))
                currentTargetCard = card;
            grid.addItem(card);
        }

        category.addContent(grid);
        return category;
    }

    private Category buildCustomCategory(String currentExclusiveTag, Set<Identifier> currentTags,
        Function<String, String> labelResolver) {
        Category category = new Category(I18n.get("enchantment.exclusive:custom"));

        List<Identifier> customTags = EnchantmentFlushAdapter
            .customExclusiveTags(context.allTypedEntries(Registries.ENCHANTMENT));
        if (customTags.isEmpty()) {
            Label fallback = new Label(I18n.get("enchantment.exclusive:custom.fallback"));
            fallback.setFont(VoxelFonts.of(VoxelFonts.Variant.REGULAR, 13));
            fallback.setTextFill(VoxelColors.ZINC_400);
            fallback.setPadding(new Insets(0, 16, 0, 16));
            category.addContent(fallback);
            return category;
        }

        ResponsiveGrid grid = new ResponsiveGrid(ResponsiveGrid.autoFit(256))
            .atMost(StudioBreakpoint.XL, ResponsiveGrid.fixed(1));

        for (Identifier tagIdent : customTags) {
            String rawTag = tagIdent.toString();
            String title = StudioText.resolve("enchantment_tag", tagIdent);

            EnchantmentTags card = buildCard(
                title, rawTag,
                Identifier.fromNamespaceAndPath(AssetEditor.MOD_ID, "icons/logo.svg"),
                tagIdent, rawTag,
                currentExclusiveTag, currentTags, labelResolver);
            if (rawTag.equals(currentExclusiveTag))
                currentTargetCard = card;
            grid.addItem(card);
        }

        category.addContent(grid);
        return category;
    }

    private EnchantmentTags buildCard(String title, String description, Identifier imageId,
        Identifier tagId, String rawTag, String currentExclusiveTag,
        Set<Identifier> currentTags, Function<String, String> labelResolver) {

        List<String> members = resolveTagMembers(tagId);
        boolean isTarget = rawTag.equals(currentExclusiveTag);
        boolean isMember = currentTags.contains(tagId);

        EnchantmentTags[] holder = { null };
        EnchantmentTags card = new EnchantmentTags(
            title, description, imageId, members, isTarget, isMember, false,
            checked -> handleTargetToggle(holder[0], tagId, checked),
            checked -> handleMembershipToggle(holder[0], tagId, checked),
            labelResolver);
        holder[0] = card;
        return card;
    }

    private void handleTargetToggle(EnchantmentTags card, Identifier tagId, boolean checked) {
        EnchantmentTags previousTarget = currentTargetCard;

        if (checked && previousTarget != null && previousTarget != card) {
            previousTarget.updateTarget(false);
        }
        card.updateTarget(checked);
        currentTargetCard = checked ? card : null;

        boolean ok = dispatchAction.apply(new EditorAction.SetExclusiveSet(checked ? tagId.toString() : ""));
        if (!ok) {
            card.updateTarget(!checked);
            if (previousTarget != null && previousTarget != card) {
                previousTarget.updateTarget(true);
            }

            currentTargetCard = previousTarget;
        }
    }

    private void handleMembershipToggle(EnchantmentTags card, Identifier tagId, boolean checked) {
        List<String> before = card.getValues();

        card.updateMembership(checked);
        List<String> after = new ArrayList<>(before);
        String enchId = elementId.toString();
        if (checked && !after.contains(enchId))
            after.add(enchId);
        if (!checked)
            after.remove(enchId);
        card.updateValues(after);

        boolean ok = applyTag.apply(tagId);

        if (!ok) {
            card.updateMembership(!checked);
            card.updateValues(before);
        }
    }

    private List<String> resolveTagMembers(Identifier tagId) {
        if (tagId == null)
            return List.of();
        List<String> members = new ArrayList<>();
        for (var entry : context.allTypedEntries(Registries.ENCHANTMENT)) {
            if (entry.tags().contains(tagId)) {
                members.add(entry.id().toString());
            }
        }
        return members;
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/components/page/enchantment/ExclusiveGroupSection.kt
package fr.hardel.asset_editor.client.compose.components.page.enchantment

import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.unit.dp
import fr.hardel.asset_editor.AssetEditor
import fr.hardel.asset_editor.client.compose.lib.rememberServerData
import fr.hardel.asset_editor.client.memory.session.server.StudioDataSlots
import fr.hardel.asset_editor.data.compendium.CompendiumTagGroup
import fr.hardel.asset_editor.client.compose.VoxelColors
import fr.hardel.asset_editor.client.compose.VoxelTypography
import fr.hardel.asset_editor.client.compose.components.ui.Category
import fr.hardel.asset_editor.client.compose.components.ui.BreakpointRule
import fr.hardel.asset_editor.client.compose.components.ui.LayoutSpec
import fr.hardel.asset_editor.client.compose.components.ui.ResponsiveGrid
import fr.hardel.asset_editor.client.compose.lib.StudioContext
import fr.hardel.asset_editor.client.compose.lib.StudioText
import fr.hardel.asset_editor.client.compose.StudioBreakpoint
import fr.hardel.asset_editor.client.compose.lib.rememberRegistryEntries
import fr.hardel.asset_editor.client.memory.session.server.ClientWorkspaceRegistries
import fr.hardel.asset_editor.workspace.flush.ElementEntry
import fr.hardel.asset_editor.workspace.flush.adapter.EnchantmentFlushAdapter
import java.util.ArrayList
import java.util.LinkedHashMap
import net.minecraft.client.resources.language.I18n
import net.minecraft.core.registries.Registries
import net.minecraft.resources.Identifier
import net.minecraft.world.item.enchantment.Enchantment

private val EXCLUSIVE_GROUP = Identifier.fromNamespaceAndPath("asset_editor", "exclusive")

private data class ExclusiveSetGroup(val tagId: Identifier) {

    fun image(): Identifier =
        tagId.withPath("textures/studio/enchantment/${tagId.path}.png")
}

@Composable
fun ExclusiveGroupSection(
    context: StudioContext,
    currentExclusiveTag: String,
    currentTags: Set<Identifier>,
    onTargetToggle: (Identifier, Boolean) -> Unit,
    onMembershipToggle: (Identifier, Boolean) -> Unit
) {
    val allEntries = rememberRegistryEntries(context, ClientWorkspaceRegistries.ENCHANTMENT)
    val membersByTag = remember(allEntries) { tagMembers(allEntries) }
    val customTags = remember(allEntries) { EnchantmentFlushAdapter.customExclusiveTags(allEntries) }
    val enchantmentGroups = rememberServerData(StudioDataSlots.COMPENDIUM_ENCHANTMENTS)
    val vanillaGroups = remember(enchantmentGroups) {
        CompendiumTagGroup.findEntries(enchantmentGroups, EXCLUSIVE_GROUP)
            .map { entry -> ExclusiveSetGroup(entry.id()) }
    }
    val labelResolver: (String) -> String = { value ->
        if (value.isBlank()) {
            ""
        } else {
            val parsed = Identifier.tryParse(value.removePrefix("#"))
            if (parsed == null) {
                value
            } else {
                context.resolveHolder(Registries.ENCHANTMENT, parsed)
                    .map { holder -> holder.value().description().string }
                    .orElseGet { StudioText.resolve(Registries.ENCHANTMENT, parsed) }
            }
        }
    }

    Category(I18n.get("enchantment.exclusive:vanilla")) {
        ResponsiveGrid(
            items = vanillaGroups.map { group ->
                {
                    EnchantmentTags(
                        title = StudioText.resolve("enchantment_tag", group.tagId),
                        description = I18n.get("enchantment_tag:${group.tagId}.desc"),
                        imageId = group.image(),
                        values = membersByTag[group.tagId].orEmpty(),
                        isTarget = currentExclusiveTag == group.tagId.toString(),
                        isMember = currentTags.contains(group.tagId),
                        locked = false,
                        onTargetToggle = { checked -> onTargetToggle(group.tagId, checked) },
                        onMembershipToggle = { checked -> onMembershipToggle(group.tagId, checked) },
                        labelResolver = labelResolver
                    )
                }
            },
            defaultSpec = LayoutSpec.AutoFit(256.dp),
            rules = listOf(BreakpointRule(maxWidth = StudioBreakpoint.XL.px.dp, spec = LayoutSpec.Fixed(floatArrayOf(1f))))
        )
    }

    Category(I18n.get("enchantment.exclusive:custom")) {
        if (customTags.isEmpty()) {
            Text(
                text = I18n.get("enchantment.exclusive:custom.fallback"),
                style = VoxelTypography.regular(13),
                color = VoxelColors.Zinc400
            )
        } else {
            ResponsiveGrid(
                items = customTags.map { tagId ->
                    {
                        EnchantmentTags(
                            title = StudioText.resolve("enchantment_tag", tagId),
                            description = tagId.toString(),
                            imageId = Identifier.fromNamespaceAndPath(AssetEditor.MOD_ID, "icons/logo.svg"),
                            values = membersByTag[tagId].orEmpty(),
                            isTarget = currentExclusiveTag == tagId.toString(),
                            isMember = currentTags.contains(tagId),
                            locked = false,
                            onTargetToggle = { checked -> onTargetToggle(tagId, checked) },
                            onMembershipToggle = { checked -> onMembershipToggle(tagId, checked) },
                            labelResolver = labelResolver
                        )
                    }
                },
                defaultSpec = LayoutSpec.AutoFit(256.dp),
                rules = listOf(BreakpointRule(maxWidth = StudioBreakpoint.XL.px.dp, spec = LayoutSpec.Fixed(floatArrayOf(1f))))
            )
        }
    }
}

private fun tagMembers(
    entries: List<ElementEntry<Enchantment>>
): Map<Identifier, List<String>> {
    val members = LinkedHashMap<Identifier, MutableList<String>>()
    entries.forEach { entry ->
        val elementId = entry.id().toString()
        entry.tags().forEach { tagId ->
            members.getOrPut(tagId) { ArrayList() }.add(elementId)
        }
    }
    return members
}
